### PR TITLE
Fix resumption trace collection and printing logic

### PR DESCRIPTION
**Pull request by eupp**
Events from coroutine resumptions were not printed before. 
Moreover, the "stack trace recovery after resumption" logic seem to be broken too.

This PR fixes these problems. 

See the output of the affected tests to spot the difference. 
One new test with several nested `suspend fun`-s was also added, to test the case of more complicated resumption logic.

### CODE REVIEW

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:224]: Let's rename `t` to `thread` for clarity?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:238]: `iThread` -> `threadId`

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:265]: Please leave a comment why `runWithoutIgnoredSection` should be here. Also, consider renaming the function to `runOutsideIgnoredSection`

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:107]: Why would you need to skip the next beforeEvent? Why should it be added now?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/LoopDetector.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/LoopDetector.kt:919]: `methodInvocation` or `methodCall`?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:294]: Consider renaming to `runOutsideIgnoredSection`

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:295]: Can this function be called outside the testing code?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:95]: why `nextId`?

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:295]: Currently, we use it only in single place (for coroutine resumption). 
Here it is guaranteed that the thread is in testing code. 
So I changed the `if` to `require`.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:95]: Renamed to `callStackTraceElementId`.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:49]: unnecessary change, let's revert it

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:1125]: Please add argument names

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:107]: This is essentially a hack to make the plugin integration working without refactoring too much code.

So the problem is because of the coroutine resumptions. 
We don't have coroutine resumptions events as trace points in the trace. Thus they don't have "eventId"-s which are used to integrate Lincheck with the plugin. 

However, when the resumption is called, a lot of stuff is happening under the hood. 
In particular, a `suspend fun` is called with the given completion object.
`MethodCallTransformer` instruments this call as it instruments other method calls, however, in case of resumption this `suspend fun` call should not create new trace point. 
`MethodCallTransformer` does not know about it, it always injects `beforeEvent` call regardless:

```
invokeStatic(Injections::beforeMethodCall)
invokeBeforeEventIfPluginEnabled("method call $methodName", setMethodEventId = true)  
```

Therefore, to "skip" this `beforeEvent` call following resumption call to `suspend fun`, we use this `skipNextBeforeEvent` flag. 


IMO, a better approach would be to refactor a code, and instead just assign `eventId`-s directly to trace points. 
Methods like `beforeMethodCall` then can return `eventId` of the created trace point, or something like `-1` in case when no trace point is created. Then subsequent `beforeEvent` call can just take this `eventId` from the stack. 

But this would be quite a large refactoring, I think it's better to implement it as separate PR. 

@avpotapov00 WDYT?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/LoopDetector.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/LoopDetector.kt:919]: I would prefer `methodInvocationId` or `methodCallId` name. 

The reason is that `methodInvocationId` is just an `Int` -- it is an identifier of method call, not a method call itself. 
The identifier is computed as hash-code of the method id and the hash-codes of all its arguments.

In case of `methodInvocation` name, I would expect the field to store the information about the method call itself, that is method name and all its arguments (as objects, not as their hash-codes).

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:295]: Oh, sorry, I was mistaken. This function can actually be called outside of testing code in case of stress strategy. Let's just leave this `if` as is.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:1149]: For consistency with other methods in `ManagedStrategy`, which usually take `iThread` as a first argument. 

So `iThread` goes as first argument, then there goes a bunch of arguments related to method call: `owner`, `className`, `methodName`, `params`; and then other arguments.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:1330]: Object whose method is called. 

Should I rename it? Would `owner` or just `obj` be better?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt
Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:107]: I'm OK with this hack, please document it in the source code.

Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt:107]: Agreed to leave this for now. Should add a comment to source code.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:1330]: Let's use `owner` or `obj`, similarly to other places in Lincheck

**Review by ndkoval — APPROVED** [review]

**Comment by ndkoval** [comment]
@eupp Approved. Should we wait for a review by @avpotapov00?

**Comment by eupp** [comment]
Yes, he said he want to have a look.

**Review by avpotapov00 — APPROVED** [review]

**Pull request merged**
